src/TableColumn/Generic.php

```diff
@@ -294,74 +286,69 @@ public function getTag($position, $value, $attr = [])
+    public function getHeaderCellHTML(Field $field = null, $value = null)
     {
-        $attr = [];
         if (!$this->table) {
-            throw new \atk4\ui\Exception(['How $table could not be set??', 'f' => $f, 'value' => $value]);
+            throw new \atk4\ui\Exception(['How $table could not be set??', 'field' => $field, 'value' => $value]);
         }
 
-        if ($tag = $this->table->hook('getColumnHeaderCell', [$this, $f, $value])) {
+        if ($tag = $this->table->hook('getColumnHeaderCell', [$this, $field, $value])) {
             return $tag[0];
         }
 
-        if ($f === null) {
+        if ($field === null) {
             return $this->getTag('head', $this->caption ?: '', $this->table->sortable ? ['class' => ['disabled']] : []);
         }
 
         // if $this->caption is empty, header caption will be overriden by linked field definition
-        $captionHtmlTag = ['div', ['class' => 'atk-table-column-header'], empty($this->caption) ? $f->getCaption() : $this->caption];
+        $caption = $this->caption ?: $field->getCaption();
+
+        $attr = [
+            'data-column' => $this->columnData
+        ];
+
+        $class = 'atk-table-column-header';
+
+        if ($this->hasHeaderAction) {
+            $attr['id'] = $this->name.'_th';
+
+            //add the action tag to the caption
+            $caption = [$caption, $this->headerActionTag];
+        }
 
         // If table is being sorted by THIS column, set the proper class
-        $attr['data-column'] = $this->columnData;
         if ($this->table->sortable) {
-            $attr['data-sort'] = $f->short_name;
-            if ($this->table->sort_by === $f->short_name) {
-                $captionHtmlTag[1]['class'] = $captionHtmlTag[1]['class'].' '.'sorted '.$this->table->sort_order;
+            $attr['data-sort'] = $field->short_name;
+            if ($this->table->sort_by === $field->short_name) {
+                $class .= ' sorted '.$this->table->sort_order;
 
                 if ($this->table->sort_order === 'ascending') {
-                    $attr['data-sort'] = '-'.$f->short_name;
+                    $attr['data-sort'] = '-'.$field->short_name;
                 } elseif ($this->table->sort_order === 'descending') {
                     $attr['data-sort'] = '';
                 }
             }
         }
 
-        if ($this->hasHeaderAction) {
-            $attr = array_merge($attr, ['id' => $this->name.'_th']);
-            $tag = $this->getTag(
-                'head',
-                [$captionHtmlTag,
-                    $this->headerActionTag,
-                ],
-                $attr
-            );
-        } else {
-            $tag = $this->getTag(
-                'head',
-                [$captionHtmlTag],
-                $attr
-            );
-        }
-
-        return $tag;
+        return $this->getTag('head', [['div', compact('class'), $caption]], $attr);
```

Comment:
I figured out how to do that but I get a problem with generating tags.

When generating the tag for the header all works well on atk4 side on the line linked to this comment.

The generated tag (var_dump) is:
```
<th data-column="accessed_at" id="atk_layoutview_fileaccesshistory_crud_view_table_generic_th" data-sort="accessed_at">
	<div class="atk-table-column-header">
		<div class="atk-table-dropdown">
			<i id="atk_layoutview_fileaccesshistory_crud_view_table_generic_ac" class="table-filter-off icon">
		</div>Accessed At</div>
</th>
```

Yet in the browser the column title is wrapped with a copy of the <i> tag?

```
<th data-column="accessed_at" id="atk_layoutview_fileaccesshistory_crud_view_table_generic_th" data-sort="-accessed_at">
	<div class="atk-table-column-header sorted ascending">
		<div class="atk-table-dropdown">
			<i id="atk_layoutview_fileaccesshistory_crud_view_table_generic_ac" class="table-filter-off icon"/>
		</div>
		<i id="atk_layoutview_fileaccesshistory_crud_view_table_generic_ac" class="table-filter-off icon">Accessed At</i>
	</div>
</th>
```

All works fine is text is "declared" first in the atk tag array

Any idea?

Reply:
Found the issue - it is related to the fact that `<i>` tag is not closed when generated by atk